File: src/verify_backward_sound.cpp
// verify_backward_sound.cpp

#include <format>
#include <iostream>

#include "test_utils.h"
#include "verify_utils.h"

std::string get_losing_name(int side_to_move, int ply)
{
  return std::format("backward,ply_max={:03d},side={:d},losing", ply, side_to_move);
}

std::string get_winning_name(int side_to_move, int ply)
{
  return std::format("backward,ply_max={:03d},side={:d},winning", ply, side_to_move);
}

void verify_backward_sound(const Game& game, int ply_max)
{
  // ply 0

  for(int side_to_move = 0; side_to_move < 2; ++side_to_move)
    {
      std::string lost_name = get_losing_name(side_to_move, 0);
      verify_lost_sound(game, side_to_move, lost_name);

      std::string won_name = get_winning_name(side_to_move, 0);
      verify_won_sound(game, side_to_move, won_name);
    }

  // later ply
  for(int ply = 1; ply <= ply_max; ++ply)
    {
      for(int side_to_move = 0; side_to_move < 2; ++side_to_move)
        {
          std::string losing_curr_name = get_losing_name(side_to_move, ply);
          std::string winning_prev_name = get_winning_name(1 - side_to_move, ply - 1);
          verify_losing_sound(game, side_to_move, losing_curr_name, winning_prev_name);

          std::string winning_curr_name = get_winning_name(side_to_move, ply);
          std::string losing_prev_name = get_losing_name(1 - side_to_move, ply - 1);
          verify_winning_sound(game, side_to_move, winning_curr_name, losing_prev_name);
        }
    }
}

int main(int argc, char **argv)
{
  if(argc < 2)
    {
      std::cerr << "usage: verify_backward_sound GAME PLY_MAX" << std::endl;
      return 1;
    }

  std::string game_name(argv[1]);
  Game *game = get_game(game_name);

  int ply_max = (argc >= 3) ? atoi(argv[2]) : 1;

  verify_backward_sound(*game, ply_max);

  return 0;
}

File: rust/dfa-games/src/bin/verify-backward-sound.rs
//! Check a whole backward solve, ply by ply.  Rust port of
//! `src/verify_backward_sound.cpp`.
//!
//! Ply 0 is the base case: the `losing` DFA holds terminal lost positions and
//! the `winning` DFA holds terminal won positions.  Every later ply is
//! relative to the previous one for the other side.

use std::path::PathBuf;
use std::process::ExitCode;

use clap::Parser;
use dfa_format::{union, Dfa};
use dfa_games::{get_game, load, verify};

#[derive(Parser, Debug)]
#[command(about = "Verify every ply of a backward solve", long_about = None)]
struct Args {
    /// Scratch directory holding <game>/ and dfas_by_hash/
    #[arg(long, default_value = "scratch")]
    scratch: PathBuf,

    /// Game to verify, e.g. breakthrough_4x4
    game: String,

    /// Highest ply to verify
    #[arg(default_value_t = 1)]
    ply_max: u32,
}

fn losing_eq_name(side_to_move: u32, ply: u32) -> String {
    format!("backward,ply={ply:03},side={side_to_move},losing")
}

fn losing_max_name(side_to_move: u32, ply: u32) -> String {
    format!("backward,ply_max={ply:03},side={side_to_move},losing")
}

fn winning_eq_name(side_to_move: u32, ply: u32) -> String {
    format!("backward,ply={ply:03},side={side_to_move},winning")
}

fn winning_max_name(side_to_move: u32, ply: u32) -> String {
    format!("backward,ply_max={ply:03},side={side_to_move},winning")
}

fn main() -> ExitCode {
    let args = Args::parse();
    match run(&args) {
        Ok(()) => ExitCode::SUCCESS,
        Err(e) => {
            eprintln!("{e:#}");
            ExitCode::FAILURE
        }
    }
}

fn run(args: &Args) -> anyhow::Result<()> {
    let game = get_game(&args.game)?;
    let game = game.as_ref();

    let open = |name: &str| -> anyhow::Result<Dfa> { load::load(&args.scratch, game, name) };

    // ply 0: the base cases, which stand on the rules alone.
    for side_to_move in 0..2 {
        let lost = losing_max_name(side_to_move, 0);
        verify::verify_lost_sound(game, side_to_move, &open(&lost)?, &lost)?;

        let won = winning_max_name(side_to_move, 0);
        verify::verify_won_sound(game, side_to_move, &open(&won)?, &won)?;
    }

    // later ply: each one against the other side's previous ply.
    for ply in 1..=args.ply_max {
        for side_to_move in 0..2 {
            let losing_curr_name = losing_max_name(side_to_move, ply);
            let losing_curr = open(&losing_curr_name)?;

            let losing_prev_name = losing_max_name(side_to_move, ply - 1);
            let losing_prev = open(&losing_prev_name)?;

            let losing_new_name = losing_eq_name(side_to_move, ply);
            let losing_new = open(&losing_new_name)?;

            union::verify_dfa_union(
                &losing_curr,
                &losing_curr_name,
                &losing_prev,
                &losing_prev_name,
                &losing_new,
                &losing_new_name,
            )?;

            let winning_prev_name = winning_max_name(1 - side_to_move, ply - 1);
            let winning_prev = open(&winning_prev_name)?;

            verify::verify_losing_sound(
                game,
                side_to_move,
                &losing_new,
                &losing_new_name,
                &winning_prev,
                &winning_prev_name,
            )?;

            let winning_curr_name = winning_max_name(side_to_move, ply);
            let winning_curr = open(&winning_curr_name)?;

            let winning_prev_name = winning_max_name(side_to_move, ply - 1);
            let winning_prev = open(&winning_prev_name)?;

            let winning_new_name = winning_eq_name(side_to_move, ply);
            let winning_new = open(&winning_new_name)?;

            union::verify_dfa_union(
                &winning_curr,
                &winning_curr_name,
                &winning_prev,
                &winning_prev_name,
                &winning_new,
                &winning_new_name,
            )?;

            let losing_prev_name = losing_max_name(1 - side_to_move, ply - 1);
            let losing_prev = open(&losing_prev_name)?;

            verify::verify_winning_sound(
                game,
                side_to_move,
                &winning_new,
                &winning_new_name,
                &losing_prev,
                &losing_prev_name,
            )?;
        }
    }

    Ok(())
}
